h/conftest.py

```diff
@@ -69,6 +69,36 @@ def settings():
     return settings
 
 
+@pytest.fixture(scope='session', autouse=True)
+def setup_database(settings):
+    """Set up the database connection and create tables."""
+    engine = db.make_engine(settings)
+    db.bind_engine(engine, should_create=True, should_drop=True)
+    api_db.use_session(db.Session)
+
+
+@pytest.fixture(autouse=True)
+def database_session(request, monkeypatch):
+    """
+    Prepare the SQLAlchemy session object.
+
+    We enable fast repeatable database tests by setting up the database only
+    once per session (see :func:`setup_database`) and then wrapping each test
+    function in a SAVEPOINT/ROLLBACK TO SAVEPOINT within the transaction.
+    """
+    savepoint = transaction.savepoint()
+
+    def _cleanup():
+        db.Session.remove()
```

Comment:
The honorable gentleman makes an excellent point. It's worth noting that `monkeypatch`es are automatically undone by a finalizer, but in this case there's no guarantee that they'll be undone in the right order. Additionally, `savepoint.rollback()` will ensure that the session is restored to the correct state anyway, so it's not necessary here.